File: Advanced Methods of Programming/JavaClientServer/Persistence/src/main/java/Repo/RepoProba.java
package Repo;

import Model.Proba;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class RepoProba implements IRepository<Integer, Proba> {
    private JdbcUtils dbUtils;

    private static final Logger logger= LogManager.getLogger();

    public RepoProba(Properties props) {
        logger.info("Initializing RepoProba with properties: {} ",props);
        this.dbUtils = new JdbcUtils(props);
    }

    public RepoProba() throws IOException {
        this.dbUtils = new JdbcUtils();
    }

    public void save(Proba proba) {
        logger.traceEntry("saving proba{}",proba);
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into proba values (?,?,?)")){
            preStmt.setInt(1,proba.getId());
            preStmt.setString(2,proba.getStil().toString());
            preStmt.setInt(3,proba.getDistanta().getValueInt(proba.getDistanta().toString()));
            preStmt.executeUpdate();
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit();
    }


    @Override
    public int size() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("select count(*) as SIZE from proba")) {
            try(ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    logger.traceExit(result.getInt("SIZE"));
                    return result.getInt("SIZE");
                }
            }
        }catch(SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        return 0;
    }

    @Override
    public Proba findOne(Integer id) {
        Connection con=dbUtils.getConnection();
        logger.traceEntry("finding proba with id {}",id);
        try(PreparedStatement preStmt=con.prepareStatement("select * from proba where idProba=?")) {
            preStmt.setInt(1,id);
            try(ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    Proba p = new Proba(result.getInt("idProba"),
                            result.getInt("distanta"),
                            result.getString("stil"));
                    logger.traceExit(p);
                    return p;
                }
            }
        }catch(SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit("No proba found with id{}",id);
        return null;
    }

    @Override
    public Iterable<Proba> findAll() {
        logger.traceEntry();
        List<Proba> lst = new ArrayList<>();
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("select * from proba")) {
            try(ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    lst.add(new Proba(result.getInt("idProba"),
                                    result.getInt("distanta"),
                            result.getString("stil")));
                }
            }
        }catch(SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit(lst);
        return lst;
    }

    public Integer numarInscrieri(Integer idProba){
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("select count(*) as NR from inscriere where idProba=?")) {
            preStmt.setInt(1,idProba);
            try(ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    logger.traceExit(result.getInt("NR"));
                    return result.getInt("NR");
                }
            }
        }catch(SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        return 0;
    }
}

File: Advanced Methods of Programming/Lab1C-MPP/Lab1C-MPP/Repo/RepoProba.cs
using System;
using System.Collections.Generic;
using System.Data;
using Lab1C_MPP.Domain;

namespace Lab1C_MPP.Repo
{
    class RepoProba : IRepository<int, Proba>
    {
        //private static readonly ILog log = LogManager.GetLogger("SortingTaskDbRepository");

        IDictionary<String, string> props;
        public RepoProba(IDictionary<String, string> props)
        {
            //log.Info("Creating RepoProba");
            this.props = props;
        }

        public IEnumerable<Proba> FindAll()
        {
            IDbConnection con = DBUtils.getConnection(props);
            IList<Proba> probe = new List<Proba>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select idProba,stil,distanta from proba";
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int idProba = dataR.GetInt32(0);
                        string stil = dataR.GetString(1);
                        int distanta = dataR.GetInt32(2);
                        Proba proba = new Proba(idProba,stil,distanta);
                        probe.Add(proba);
                    }
                }
            }

            return probe;
        }

        public Proba FindOne(int Id)
        {
            //log.InfoFormat("Entering findOne with value {0}", id);
            IDbConnection con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select idProba,stil,distanta from proba where idProba=@idProba";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@idProba";
                paramId.Value = Id;
                comm.Parameters.Add(paramId);
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int idProba = dataR.GetInt32(0);
                        string stil = dataR.GetString(1);
                        int distanta = dataR.GetInt32(2);
                        Proba proba = new Proba(idProba, stil, distanta);
                        //log.InfoFormat("Exiting findOne with value {0}", task);
                        return proba;
                    }
                }
            }
            //log.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }

        public void Save(Proba proba)
        {
            var con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into proba  values (@idProba,@stil,@distanta)";
                var paramId = comm.CreateParameter();
                paramId.ParameterName = "@idProba";
                paramId.Value = proba.GetId();
                comm.Parameters.Add(paramId);

                var paramStil = comm.CreateParameter();
                paramStil.ParameterName = "@stil";
                paramStil.Value =proba.GetStil();
                comm.Parameters.Add(paramStil);

                var paramDistanta = comm.CreateParameter();
                paramDistanta.ParameterName = "@distanta";
                paramDistanta.Value = proba.GetDistanta();
                comm.Parameters.Add(paramDistanta);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("Nu s-a adaugat niciun proba!");
            }
        }

        public int Size()
        {
            IDbConnection con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select count(*) from proba";
                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        //log.InfoFormat("Exiting findOne with value {0}", task);
                        return dataR.GetInt32(0);
                    }
                }
            }
            //log.InfoFormat("Exiting findOne with value {0}", null);
            return 0;
        }

        public int NumarInscrieri(int idProba)
        {
            IDbConnection con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select count(*) from inscriere where idProba=@idProba ";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@idProba";
                paramId.Value = idProba;
                comm.Parameters.Add(paramId);
                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        //log.InfoFormat("Exiting findOne with value {0}", task);
                        return dataR.GetInt32(0);
                    }
                }
            }
            //log.InfoFormat("Exiting findOne with value {0}", null);
            return 0;
        }
    }
}
